cpp/tests/experimental/trajectory/trajectory_test_utils.cuh

```diff
@@ -0,0 +1,190 @@
+#include <thrust/tabulate.h>
+#include <thrust/transform.h>
+
+#include <chrono>
+#include <cstdint>
+
+namespace cuspatial {
+namespace test {
+
+using time_point = std::chrono::time_point<std::chrono::system_clock, std::chrono::milliseconds>;
+
+/* Test data generation for trajectory APIs. Generates num_trajectories trajectories of random
+   size between 1 and max_trajectory_size points. Creates both a reference (sorted) set of
+   trajectory IDs, timestamps, and points, and a shuffled (unsorted) set of the same for test
+   input. The sorted data can be input directly into `trajectory_distance_and_speed`, while
+   the shuffled data can be input to `derive_trajectories`.
+
+   The times are not random, but do vary somewhat between trajectories and trajectory lenghts.
+
+   Likewise, the positions are not random, but follow a sinusoid pattern based on the time stamps.
+*/
+template <typename T>
+struct trajectory_test_data {
+  rmm::device_vector<std::int32_t> offsets;
+
+  rmm::device_vector<std::int32_t> ids;
+  rmm::device_vector<time_point> times;
+  rmm::device_vector<cuspatial::vec_2d<T>> points;
+
+  rmm::device_vector<std::int32_t> ids_sorted;
+  rmm::device_vector<time_point> times_sorted;
+  rmm::device_vector<cuspatial::vec_2d<T>> points_sorted;
+
+  trajectory_test_data(std::size_t num_trajectories, std::size_t max_trajectory_size)
+  {
+    make_data(num_trajectories, max_trajectory_size);
+  }
+
+  struct id_functor {
+    std::int32_t* offsets_begin{};
+    std::int32_t* offsets_end{};
+
+    id_functor(std::int32_t* offsets_begin, std::int32_t* offsets_end)
+      : offsets_begin(offsets_begin), offsets_end(offsets_end)
+    {
+    }
+
+    __device__ std::int32_t operator()(int i)
+    {
+      // Find the index within the current trajectory
+      return thrust::distance(
+        offsets_begin,
+        thrust::prev(thrust::upper_bound(thrust::seq, offsets_begin, offsets_end, i)));
+    }
+  };
+
+  struct timestamp_functor {
+    std::int32_t* offsets_begin{};
+    std::int32_t* offsets_end{};
+
+    timestamp_functor(std::int32_t* offsets_begin, std::int32_t* offsets_end)
+      : offsets_begin(offsets_begin), offsets_end(offsets_end)
+    {
+    }
+
+    __device__ time_point operator()(int i, int id)
+    {
+      auto offset = thrust::prev(thrust::upper_bound(thrust::seq, offsets_begin, offsets_end, i));
+      auto time_step = i - *offset;
+      // The arithmetic here just adds some variance to the time step but keeps it monotonically
+      // increasing with `i`
+      auto duration = (id % 10) * std::chrono::milliseconds(1000) +
+                      time_step * std::chrono::milliseconds(100) +
+                      std::chrono::milliseconds(int(10 * cos(time_step)));
+      return time_point{duration};
+    }
+  };
+
+  struct point_functor {
+    __device__ cuspatial::vec_2d<T> operator()(time_point const& time, std::int32_t id)
+    {
+      // X is time in seconds, Y is cosine(time), offset by ID
```

Comment:
I think `count` only counts the ticks of the specified time unit: https://en.cppreference.com/w/cpp/chrono/duration/count
```suggestion
      // X is time in milliseconds, Y is cosine(time), offset by ID
```